src/NewSkin.cpp

```diff
@@ -1100,3 +1161,197 @@ bool NewSkinLoader::load_into_data(StepsType stype,
+		int field_default= lua_gettop(L);
+		int field_key= field_default - 1;
+#define SAFE_CONTINUE lua_settop(L, field_key); continue;
+		int key_type= lua_type(L, field_key);
+		int field_type= lua_type(L, field_default);
+		lua_pushvalue(L, field_key);
+		lua_gettable(L, curr_param_set_info);
+		int field_info= lua_gettop(L);
+		if(lua_type(L, field_info) != LUA_TTABLE)
+		{
+			// Push a copy of the key onto the stack so that lua_tostring
+			// won't affect the key being used to iterate through the table.
+			lua_pushvalue(L, field_key);
+			LuaHelpers::ReportScriptErrorFmt("Field %s in noteskin parameters has invalid type info.", lua_tostring(L, -1));
+			SAFE_CONTINUE;
+		}
+		// Ignore stuff that doesn't have a string or number index.
+		if(key_type != LUA_TSTRING && key_type != LUA_TNUMBER)
+		{
+			SAFE_CONTINUE;
+		}
+		switch(field_type)
+		{
+			case LUA_TTABLE:
+				{
+					void const* table= lua_topointer(L, field_default);
+					bool visited= (visited_tables.find(table) != visited_tables.end());
+					if(visited)
+					{
+						SAFE_CONTINUE;
+					}
+					if(curr_depth >= max_depth)
+					{
+						LuaHelpers::ReportScriptError("Noteskin parameter table exceeded maximum depth, ignoring deeper parameters.");
+						SAFE_CONTINUE;
+					}
+					lua_pushvalue(L, field_key); // stack: field_key, field_default, field_info, field_key
+					lua_gettable(L, curr_param_set_dest); // stack: field_key, field_default, field_info, next_param_set_dest
+					int next_param_set_dest= lua_gettop(L);
+					if(lua_type(L, next_param_set_dest) != LUA_TTABLE)
+					{
+						lua_pop(L, 1); // stack: field_key, field_default, field_info
+						lua_pushvalue(L, field_key); // stack: field_key, field_default, field_info, field_key
+						lua_newtable(L); // stack: field_key, field_default, field_info, field_key, table
+						lua_settable(L, curr_param_set_dest); // stack: field_key, field_default, field_info
+						lua_pushvalue(L, field_key); // stack: field_key, field_default, field_info, field_key
+						lua_gettable(L, curr_param_set_dest); // stack: field_key, field_default, field_info, next_param_set_dest
+						next_param_set_dest= lua_gettop(L);
+					}
+					visited_tables.insert(table);
+					recursive_sanitize_skin_parameters(L, visited_tables, curr_depth+1,
+						field_info, field_default, next_param_set_dest);
+					// Pop next_param_set_dest.
+					lua_pop(L, 1); // stack: field_key, field_default, field_info
+				}
+				break;
+			case LUA_TSTRING:
+			case LUA_TNUMBER:
+			case LUA_TBOOLEAN:
+				{
+#define SET_FIELD_TO_DEFAULT \
+	lua_pushvalue(L, field_key); \
+	lua_pushvalue(L, field_default); \
+	lua_settable(L, curr_param_set_dest);
+					lua_pushvalue(L, field_key);
+					lua_gettable(L, curr_param_set_dest);
+					int dest_value= lua_gettop(L);
+					if(lua_type(L, dest_value) != field_type)
+					{
+						SET_FIELD_TO_DEFAULT;
+					}
+					else if(field_type != LUA_TBOOLEAN)
+					{
+						lua_getfield(L, field_info, "values");
```

Comment:
field info in noteskin.lua has "choices" table, but "values" is used in this. either is correct?